comprl/server/matchmaking.py

```diff
@@ -1,23 +1,119 @@
 """contains matchmaking logic"""
 
-from queue import Queue
+from datetime import datetime
+from openskill.models import PlackettLuce
+from typing import TypeAlias
 
-from . import game_manager
+
+from . import game_manager, user_database
 
 from comprl.shared.types import PlayerID
 
-_queue: Queue[PlayerID] = Queue()
+# from .player_manager import get_num_online_players
+# TODO fix circular imports
+
+# Type of a player entry in the queue, containing the player ID, user ID, mu, sigma
+# and time they joined the queue
+QueuePlayer: TypeAlias = tuple[PlayerID, int, float, float, datetime]
+# queue storing player id, mu, sigma and time they joined the queue
+_queue: list[QueuePlayer] = []
+# The model used for matchmaking
+model = PlackettLuce()
+_MATCH_QUALITY_THRESHOLD = 0.8
 
 
-def match(id: PlayerID) -> None:
-    """handles mathcmaking between players and starts the game
+def match(player_id: PlayerID, user_id: int) -> None:
+    """handles matchmaking between players and starts the game
 
     Args:
         id (PlayerID): id of the player to match
+        user_ID (int): user id of the player
+        mu (float): current mu value of the player
+        sigma (float): current sigma value of the player
     """
-    # some simple quick and dirty "matchmaking"
-    if not _queue.empty():
-        game_manager.start_game([_queue.get(), id])
+
+    mu, sigma = user_database.get_matchmaking_parameters(user_id)
+
+    # check if enough players are waiting
+    if len(_queue) < _min_players_waiting():
+        _queue.append((player_id, user_id, mu, sigma, datetime.now()))
+        return
+    else:
+        now = datetime.now()
+        for player2 in _queue:
+            player1 = player_id, user_id, mu, sigma, now
+            # try matching the new player against all players in the queue
+            if _try_matching(player1, player2):
+                return
+        # no suitable match found
+        _queue.append((player_id, user_id, mu, sigma, datetime.now()))
         return
 
-    _queue.put(id)
+
+def _min_players_waiting() -> int:
+    """calculates how many players at minimum need to wait in the queue"""
+    active_players = 0  # TODO get_num_online_players()
```

Comment:
I can also not think of a good way to do this. We will have to discuss this later.

Reply:
I think we should wait on #71. Then this is possible. So I suggest closing this PR and I will open a new one, covering this issue when #71 is merged. Do you agree?